cpp/src/arrow/compute/exec/asof_join_benchmark.cc

```diff
@@ -0,0 +1,1023 @@
+        std::make_shared<arrow::fs::LocalFileSystem>();
+    arrow::compute::ExecNode* left_table_source;
+    int64_t left_table_rows;
+    int left_table_batches;
+    size_t left_table_bytes;
+    tie(left_table_source, left_table_rows, left_table_batches, left_table_bytes) =
+        make_arrow_ipc_reader_node(plan, fs, data_directory + left_table);
+    std::vector<ExecNode*> inputs = {left_table_source};
+    int right_hand_rows = 0;
+    int64_t right_hand_bytes = 0;
+    for (std::string right_table : right_tables) {
+      arrow::compute::ExecNode* right_table_source;
+      int64_t right_table_rows;
+      int right_table_batches;
+      size_t right_table_bytes;
+      tie(right_table_source, right_table_rows, right_table_batches, right_table_bytes) =
+          make_arrow_ipc_reader_node(plan, fs, data_directory + right_table);
+      inputs.push_back(right_table_source);
+      right_hand_rows += right_table_rows;
+      right_hand_bytes += right_table_bytes;
+    }
+    rows = left_table_rows + right_hand_rows;
+    bytes = left_table_bytes + right_hand_bytes;
+    ASSERT_OK_AND_ASSIGN(arrow::compute::ExecNode * asof_join_node,
+                         MakeExecNode(factory_name, plan.get(), inputs, options));
+
+    AsyncGenerator<util::optional<ExecBatch>> sink_gen;
+    MakeExecNode("sink", plan.get(), {asof_join_node}, SinkNodeOptions{&sink_gen});
+    state.ResumeTiming();
+    // std::cout << "starting and collecting with " <<
+    // ctx.memory_pool()->bytes_allocated() << std::endl;
+    ASSERT_FINISHES_OK(StartAndCollect(plan.get(), sink_gen));
+    // std::cout << "finishing with " << ctx.memory_pool()->bytes_allocated() <<
+    // std::endl;
+  }
+  // std::cout << "reporting with " << ctx.memory_pool()->bytes_allocated() << std::endl;
+  state.counters["total_rows_per_second"] = benchmark::Counter(
+      static_cast<double>(state.iterations() * rows), benchmark::Counter::kIsRate);
+
+  state.counters["total_bytes_per_second"] = benchmark::Counter(
+      static_cast<double>(state.iterations() * bytes), benchmark::Counter::kIsRate);
+}
+
+static void AsOfJoinOverhead(benchmark::State& state, std::string left_table,
+                             std::vector<std::string> right_tables) {
+  int64_t tolerance = 0;
+  AsofJoinNodeOptions options = AsofJoinNodeOptions(time_col, key_col, tolerance);
+  TableJoinOverhead(state, left_table, right_tables, "asofjoin", options);
+}
+
+static void HashJoinOverhead(benchmark::State& state, std::string left_table,
+                             std::vector<std::string> right_tables) {
+  HashJoinNodeOptions options =
+      HashJoinNodeOptions({time_col, key_col}, {time_col, key_col});
+  TableJoinOverhead(state, left_table, right_tables, "hashjoin", options);
+}
+
+// this generates the set of right hand tables to test on.
+void SetArgs(benchmark::internal::Benchmark* bench) { bench->UseRealTime(); }
+
+BENCHMARK_CAPTURE(AsOfJoinOverhead,
```

Comment:
This is a lot of possible combinations.  How long does it take to run the full suite?  Are all of these combinations informative?

Reply:
Yes, I agree -- I created a larger set just for exploratory purposes and have been refining it down. We found the most interesting results varying key density and time frequency (we observe some piecewise linear and constant relationships between real_time and the respective property, for example). 

The rest of the properties like column width, multi-table joins, and batch size seem more or less straightforward / linear -- so we can certainly reduce the amount of datapoints here.

Keep in mind we are also benchmarking hashjoin here as well for comparisons, which @icexelloss mentioned might not be suitable for this file here. Do you have any advice if those hashjoin comparison benchmarks are worth keeping in arrow?

Reply:
I agree it is initially interesting.  Long term I think the main advantage for an asof join over a hash join is going to be memory usage though.  Asof join can operate in a streaming fashion while hash join will require eventually spilling to disk as the input size increases.  So a performance benchmark between the two probably isn't needed long term.

Reply:
Yep, agree. In its current state, (still containing both hashjoin / asofjoin), it runs locally in about ~6 minutes on my laptop.